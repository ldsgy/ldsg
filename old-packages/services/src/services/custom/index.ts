import { Service, SettingsSchema } from "@ldsg/core";
import _ from "lodash";
import { CUSTOM_SERVICE_DEFAULT_SETTINGS_SCHEMA } from "./constants";

export * from "./constants";

export class CustomService extends Service {
  settingsSchema: SettingsSchema = CUSTOM_SERVICE_DEFAULT_SETTINGS_SCHEMA;

  getAlias = () => {
    const settings = this.getSettings();

    const alias = settings.alias;

    return alias;
  };

  getCamelCaseAlias = () => {
    const alias = this.getAlias();

    const camelCaseAlias = _.camelCase(alias);

    return camelCaseAlias;
  };

  appendToContext = (context: any) => {
    const camelCaseAlias = this.getCamelCaseAlias();

    const handler = this.getChildHandlerServiceHandler();

    context[camelCaseAlias] = async (params: any) => {
      if (!handler) {
        return;
      }

      return await handler(params, context);
    };
  };
}
