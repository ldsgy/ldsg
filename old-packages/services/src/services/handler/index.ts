import { Service, SettingsSchema } from "@ldsg/core";
import { HANDLER_SERVICE_DEFAULT_SETTINGS_SCHEMA } from "./constants";

export * from "./constants";

type Handler = (...rest: any[]) => any | Promise<any>;

type GetHandler = () => Handler;

export interface HandlerInfo {
  requireKey: string;
}

export class HandlerService extends Service {
  settingsSchema: SettingsSchema = HANDLER_SERVICE_DEFAULT_SETTINGS_SCHEMA;

  getHandler: GetHandler = () => {
    const moduleId = this.getModuleId();

    try {
      const requireResult = require(moduleId);

      if ("handler" in requireResult) {
        return requireResult.handler;
      } else {
        throw new Error("invalid handler");
      }
    } catch (error) {
      console.error(error);

      throw new Error("invalid require");
    }
  };

  getModuleId = (): string => {
    const settings = this.getSettings();

    const { moduleId } = settings;

    const res: string = moduleId;

    return res;
  };
}
