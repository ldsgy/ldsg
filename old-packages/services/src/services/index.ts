export * from "./application";
export * from "./custom";
export * from "./database";
export * from "./field";
export * from "./handler";
export * from "./http";
export * from "./key-value-storage";
export * from "./middleware";
export * from "./table";
