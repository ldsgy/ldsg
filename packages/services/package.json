{
  "name": "@ldsg/services",
  "version": "0.1.0-alpha.9",
  "main": "dist/index.js",
  "files": [
    "dist/"
  ],
  "types": "dist",
  "scripts": {
    "build": "rm -rf dist && tsc --project tsconfig.build.json"
  },
  "author": "wanxger",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "description": "",
  "dependencies": {
    "@formily/json-schema": "^2.3.1",
    "@ldsg/common": "workspace:*",
    "@ldsg/core": "workspace:*",
    "ali-oss": "^6.20.0",
    "graphql": "^16.8.1",
    "graphql-compose": "^9.0.10",
    "graphql-compose-mongoose": "^10.0.0",
    "graphql-middleware": "^6.1.35",
    "graphql-shield": "^7.6.5",
    "graphql-yoga": "^5.1.1",
    "ioredis": "^5.3.2",
    "json-schema": "^0.4.0",
    "lodash": "^4.17.21",
    "luxon": "^3.4.4",
    "mongodb": "^6.3.0",
    "mongoose": "^8.3.1",
    "sofa-api": "^0.18.2",
    "uri-js": "^4.4.1",
    "uuid": "^9.0.1"
  },
  "devDependencies": {
    "@ant-design/pro-components": "^2.6.43",
    "@types/ali-oss": "^6.16.11",
    "@types/bcryptjs": "^2.4.6",
    "@types/express": "^4.17.17",
    "@types/json-schema": "^7.0.15",
    "@types/lodash": "^4.14.202",
    "@types/luxon": "^3.4.2",
    "@types/node": "^20.13.0",
    "@types/recursive-readdir": "^2.2.4",
    "@types/supertest": "^6.0.0",
    "@types/uuid": "^9.0.7",
    "@typescript-eslint/eslint-plugin": "^6.0.0",
    "@typescript-eslint/parser": "^6.0.0",
    "eslint": "^8.42.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-prettier": "^5.0.0",
    "prettier": "^3.0.0",
    "recursive-readdir": "^2.2.3",
    "simple-git": "^3.24.0",
    "source-map-support": "^0.5.21",
    "supertest": "^6.3.4",
    "ts-node": "^10.9.2",
    "tsconfig-paths": "^4.2.0",
    "tslib": "^2.6.2",
    "typescript": "^5.4.5"
  }
}
