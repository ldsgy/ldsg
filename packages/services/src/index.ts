export * from "./constants";
export * from "./enum";
export * from "./services";
export * from "./types";
export * from "./utils";
