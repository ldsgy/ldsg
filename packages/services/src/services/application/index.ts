import { Service, ServiceLevel, SettingsSchema } from "@ldsg/core";
import { RequestHandler } from "express";
import { GraphQLJSONObject, SchemaComposer } from "graphql-compose";
import { applyMiddleware } from "graphql-middleware";
import { createYoga, YogaServerOptions } from "graphql-yoga";
import _ from "lodash";
import { useSofa } from "sofa-api";
import { ServiceType } from "../../enum";
import { ModifyGraphQLSchema } from "../../types";
import { APPLICATION_SERVICE_DEFAULT_SETTINGS_SCHEMA } from "./constants";
import { ApplicationContext, ApplicationInstance } from "./types";

export * from "./constants";
export * from "./types";

export interface GetApplicationInstanceParams {
  basePath?: string;
  yogaServerOptions?: YogaServerOptions<any, any>;
}

type GetApplicationInstance = (
  params?: GetApplicationInstanceParams
) => ApplicationInstance;

interface ServicesParams {
  services: Service[];
}

export class ApplicationService extends Service {
  settingsSchema: SettingsSchema = APPLICATION_SERVICE_DEFAULT_SETTINGS_SCHEMA;

  appendToContext = (context: any) => {
    context.getService = this.getService;

    context.getServices = this.getServices;
  };

  getGraphQLSchema = (params: ServicesParams) => {
    const { services } = params;

    const schemaComposer = new SchemaComposer();

    services.forEach((value) => {
      value.modifyGraphQLSchema?.({ schemaComposer });
    });

    let schema = schemaComposer.buildSchema();

    const middlewareServices = this.getServices({
      level: ServiceLevel.CHILD,
      type: ServiceType.MIDDLEWARE,
    });

    const middlewares = middlewareServices
      .sort((a, b) => a.getSettings().sort - b.getSettings().sort)
      .map((value) => value.getMiddleware());

    const filterMiddlewares = middlewares.filter((value) => {
      return !!value;
    });

    if (middlewares.length) {
      schema = applyMiddleware(schema, ...filterMiddlewares);
    }

    return schema;
  };

  getContext = (params: ServicesParams) => {
    const { services } = params;

    const context: ApplicationContext = {};

    /**
     * append context
     */
    services.forEach((value) => {
      value.appendToContext?.(context);
    });

    return context;
  };

  getApplicationInstance: GetApplicationInstance = (params) => {
    const { basePath: paramsBasePath = "/", yogaServerOptions } = params || {};

    const basePath = _.endsWith(paramsBasePath, "/")
      ? paramsBasePath.slice(0, -1)
      : paramsBasePath;

    const services = this.getServices();

    const context = this.getContext({ services });

    const schema = this.getGraphQLSchema({ services });

    /**
     * yoga
     */
    const graphqlEndpoint = `${basePath}/graphql`;

    const yoga = createYoga({
      context,
      graphqlEndpoint,
      landingPage: false,
      schema,
      ...yogaServerOptions,
    });

    /**
     * sofa
     */
    const sofaBasePath = `${basePath}/api`;

    const sofaServerUrl = sofaBasePath;

    const sofa = useSofa({
      basePath: sofaBasePath,
      schema,
      openAPI: {
        info: {
          title: "Example API",
          description: "Example API Description",
          version: "3.0.0",
        },
        servers: [{ url: sofaServerUrl }],
      },
    });

    const httpServices = this.getServices({
      level: ServiceLevel.CHILD,
      type: ServiceType.HTTP,
    });

    const otherPathRequestHandler: Record<string, RequestHandler> = {};

    httpServices.forEach((httpService) => {
      const { path } = httpService.getSettings();

      otherPathRequestHandler[path] = httpService.handleRequest;
    });

    const applicationInstance: ApplicationInstance = {
      context,
      yoga,
      sofa,
      otherPathRequestHandler,
    };

    return applicationInstance;
  };

  modifyGraphQLSchema: ModifyGraphQLSchema = (params) => {
    const { schemaComposer } = params;

    schemaComposer.Query.addFields({
      hello: {
        description: "示例，正常时返回 Hello world!",
        type: "String!",
        resolve: () => {
          return "Hello world!";
        },
      },
    });

    const ServiceTC = schemaComposer.createObjectTC({
      name: "Service",
      description: "服务",
      fields: {
        id: {
          description: "ID",
          type: "ID!",
        },
        property: {
          description: "属性",
          type: "JSON",
          args: {
            name: {
              description: "名称",
              type: "String!",
            },
            params: {
              description: "参数",
              type: GraphQLJSONObject,
            },
          },
          resolve: (source, args, context, info) => {
            return source.getProperty(args);
          },
        },
      },
    });

    const settings = this.getSettings();

    const { exposeServiceInfo = false } = settings;

    if (exposeServiceInfo) {
      schemaComposer.Query.addFields({
        service: {
          description: "服务",
          type: ServiceTC.NonNull,
          args: {
            params: {
              description: "参数",
              type: GraphQLJSONObject,
            },
          },
          resolve: (source, args, context, info) => {
            const { params = {} } = args;

            return this.getService(params);
          },
        },
        services: {
          description: "服务组",
          type: ServiceTC.NonNull.List.NonNull,
          args: {
            params: {
              description: "参数",
              type: GraphQLJSONObject,
            },
          },
          resolve: (source, args, context, info) => {
            const { params = {} } = args;

            return this.getServices(params);
          },
        },
      });
    }
  };
}
