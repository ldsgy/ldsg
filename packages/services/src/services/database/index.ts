import { Service, ServiceLevel, SettingsSchema } from "@ldsg/core";
import mongoose from "mongoose";
import { DATABASE_SERVICE_DEFAULT_SETTINGS_SCHEMA } from "./constants";
import {
  GetMongooseConnection,
  GetMongooseCreateConnectionParams,
  MongooseCreateConnectionParams,
} from "./types";

export * from "./constants";

export class DatabaseService extends Service {
  settingsSchema: SettingsSchema = DATABASE_SERVICE_DEFAULT_SETTINGS_SCHEMA;

  appendToContext = (context: any) => {
    context.models = this.getMongooseConnectionModels();
  };

  getMongooseCreateConnectionParams: GetMongooseCreateConnectionParams = () => {
    const settings = this.getSettings();

    const { DATABASE_URI = "127.0.0.1", DATABASE_NAME = "27017" } = process.env;

    const { uri = DATABASE_URI, dbName = DATABASE_NAME } = settings;

    const result: MongooseCreateConnectionParams = {
      uri,
      options: {
        dbName,
      },
    };

    return result;
  };

  getMongooseConnection: GetMongooseConnection = () => {
    const { uri, options } = this.getMongooseCreateConnectionParams();

    const connection = mongoose.createConnection(uri, options);

    return connection;
  };

  getMongooseConnectionModels = () => {
    const connection = this.getMongooseConnection();

    const peerServices = this.getServices({
      level: ServiceLevel.PEER,
    });

    peerServices.forEach((value) =>
      value.addMongooseModel?.({
        connection,
      })
    );

    return connection.models;
  };
}
