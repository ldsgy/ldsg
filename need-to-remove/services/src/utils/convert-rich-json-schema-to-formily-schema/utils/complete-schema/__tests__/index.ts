import { completeSchema } from "..";
import { DEFAULT_TYPE_TO_DEFAULT_MAP } from "../..";
import { FIELD_SERVICE_DEFAULT_SETTINGS_SCHEMA } from "../../../../../services";

test("empty object", async () => {
  const { formilySchema } = completeSchema({
    richJsonSchema: {
      type: "object",
    },
    typeToDefaultMap: DEFAULT_TYPE_TO_DEFAULT_MAP,
  });

  expect(formilySchema).toMatchSnapshot();
});

test("field service default settings schema", async () => {
  const { formilySchema } = completeSchema({
    richJsonSchema: FIELD_SERVICE_DEFAULT_SETTINGS_SCHEMA,
    typeToDefaultMap: DEFAULT_TYPE_TO_DEFAULT_MAP,
  });

  expect(formilySchema).toMatchSnapshot();
});
