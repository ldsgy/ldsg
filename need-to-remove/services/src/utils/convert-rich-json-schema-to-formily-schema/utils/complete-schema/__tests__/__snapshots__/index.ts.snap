// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty object 1`] = `
{
  "properties": {},
  "type": "object",
}
`;

exports[`field service default settings schema 1`] = `
{
  "properties": {
    "alias": {
      "description": "用于 GraphQL 内生成的字段名，只包含字母、数字和下划线，且以字母开头",
      "pattern": "^[a-zA-Z][a-zA-Z0-9_]*$",
      "required": true,
      "title": "别名",
      "type": "string",
      "x-component": "Input",
      "x-decorator": "FormItem",
    },
    "description": {
      "description": "用于注释",
      "title": "描述",
      "type": "string",
      "x-component": "Input",
      "x-decorator": "FormItem",
    },
    "fieldsDirection": {
      "description": "用于 GraphQL，仅表单中有该字段",
      "enum": [
        {
          "label": "输入",
          "value": "INPUT",
        },
        {
          "label": "输出",
          "value": "OUTPUT",
        },
      ],
      "options": [
        {
          "label": "输入",
          "value": "INPUT",
        },
        {
          "label": "输出",
          "value": "OUTPUT",
        },
      ],
      "title": "字段组方向",
      "type": "string",
      "x-component": "Select",
      "x-decorator": "FormItem",
    },
    "isInputArgument": {
      "default": false,
      "description": "当是输入参数时，会出现在查询参数中，仅 GraphQL 中有该字段，与 objectTypeName 二选一",
      "title": "是否是输入参数",
      "type": "boolean",
      "x-component": "Checkbox",
      "x-decorator": "FormItem",
    },
    "multiple": {
      "default": false,
      "title": "多选",
      "type": "boolean",
      "x-reactions": {
        "dependencies": [
          "type",
        ],
        "fulfill": {
          "state": {
            "display": "{{["IMAGE","OPTIONS"].includes($deps[0])?"visible": "hidden"}}",
          },
        },
      },
    },
    "name": {
      "description": "具有辨识度的名称",
      "required": true,
      "title": "名称",
      "type": "string",
      "x-component": "Input",
      "x-decorator": "FormItem",
    },
    "objectTypeName": {
      "description": "此表单字段会添加至此对象类型中，仅 GraphQL 中有该字段，与 isInputArgument 二选一",
      "title": "对象类型名称",
      "type": "string",
      "x-component": "Input",
      "x-decorator": "FormItem",
    },
    "options": {
      "items": {
        "properties": {
          "space": {
            "properties": {
              "label": {
                "required": true,
                "title": "选项名",
                "type": "string",
                "x-component": "Input",
                "x-decorator": "FormItem",
              },
              "remove": {
                "type": "void",
                "x-component": "ArrayItems.Remove",
                "x-decorator": "FormItem",
              },
              "sort": {
                "type": "void",
                "x-component": "ArrayItems.SortHandle",
                "x-decorator": "FormItem",
              },
              "value": {
                "required": true,
                "title": "选项值",
                "type": "string",
                "x-component": "Input",
                "x-decorator": "FormItem",
              },
            },
            "type": "void",
            "x-component": "Space",
          },
        },
        "type": "object",
      },
      "properties": {
        "add": {
          "title": "添加",
          "type": "void",
          "x-component": "ArrayItems.Addition",
        },
      },
      "title": "选项",
      "type": "array",
      "x-component": "ArrayItems",
      "x-decorator": "FormItem",
      "x-reactions": {
        "dependencies": [
          "type",
        ],
        "fulfill": {
          "state": {
            "display": "{{$deps[0]==="OPTIONS"?"visible": "hidden"}}",
          },
        },
      },
    },
    "path": {
      "description": "用于存储，仅表格中有该字段，只包含字母、数字和下划线，且以字母开头",
      "pattern": "^[a-zA-Z][a-zA-Z0-9_]*$",
      "title": "路径",
      "type": "string",
      "x-component": "Input",
      "x-decorator": "FormItem",
    },
    "required": {
      "default": false,
      "title": "必填",
      "type": "boolean",
      "x-component": "Checkbox",
      "x-decorator": "FormItem",
    },
    "type": {
      "default": "TEXT",
      "description": "用于字段类型",
      "enum": [
        {
          "label": "文本",
          "value": "TEXT",
        },
        {
          "label": "数字",
          "value": "NUMBER",
        },
        {
          "label": "ID",
          "value": "ID",
        },
        {
          "label": "MongoID",
          "value": "MongoID",
        },
        {
          "label": "日期",
          "value": "DATE",
        },
        {
          "label": "选项",
          "value": "OPTIONS",
        },
        {
          "label": "是否",
          "value": "BOOLEAN",
        },
        {
          "label": "图片",
          "value": "IMAGE",
        },
        {
          "label": "附件",
          "value": "FILE",
        },
        {
          "label": "JSON",
          "value": "JSON",
        },
        {
          "label": "对象",
          "value": "OBJECT",
        },
        {
          "label": "列表",
          "value": "LIST",
        },
      ],
      "options": [
        {
          "label": "文本",
          "value": "TEXT",
        },
        {
          "label": "数字",
          "value": "NUMBER",
        },
        {
          "label": "ID",
          "value": "ID",
        },
        {
          "label": "MongoID",
          "value": "MongoID",
        },
        {
          "label": "日期",
          "value": "DATE",
        },
        {
          "label": "选项",
          "value": "OPTIONS",
        },
        {
          "label": "是否",
          "value": "BOOLEAN",
        },
        {
          "label": "图片",
          "value": "IMAGE",
        },
        {
          "label": "附件",
          "value": "FILE",
        },
        {
          "label": "JSON",
          "value": "JSON",
        },
        {
          "label": "对象",
          "value": "OBJECT",
        },
        {
          "label": "列表",
          "value": "LIST",
        },
      ],
      "required": true,
      "title": "类型",
      "type": "string",
      "x-component": "Select",
      "x-decorator": "FormItem",
    },
    "unique": {
      "default": false,
      "title": "唯一",
      "type": "boolean",
      "x-component": "Checkbox",
      "x-decorator": "FormItem",
    },
  },
  "required": [
    "name",
    "type",
    "alias",
  ],
  "type": "object",
}
`;
